Source/CarthageKit/Project.swift

```diff
@@ -453,23 +458,25 @@ public final class Project {
 						.flatMap(.concat, transform: unzip(archive:))
 						.flatMap(.concat) { directoryURL in
 							return frameworksInDirectory(directoryURL)
+								.flatMap(.merge) { url in
+									return checkFrameworkCompatibility(url)
+										.mapError { error in CarthageError.internalError(description: error.description) }
+								}
 								.flatMap(.merge, transform: self.copyFrameworkToBuildFolder)
 								.flatMap(.merge) { frameworkURL in
 									return self.copyDSYMToBuildFolderForFramework(frameworkURL, fromDirectoryURL: directoryURL)
 										.then(self.copyBCSymbolMapsToBuildFolderForFramework(frameworkURL, fromDirectoryURL: directoryURL))
 								}
 								.on(completed: {
 									_ = try? FileManager.default.trashItem(at: checkoutDirectoryURL, resultingItemURL: nil)
+								}, disposed: {
+									_ = try? FileManager.default.removeItem(at: directoryURL)
 								})
-								.then(SignalProducer<URL, CarthageError>(value: directoryURL))
 						}
-						.attemptMap { (temporaryDirectoryURL: URL) -> Result<Bool, CarthageError> in
-							do {
-								try FileManager.default.removeItem(at: temporaryDirectoryURL)
-								return .success(true)
-							} catch let error as NSError {
-								return .failure(.writeFailed(temporaryDirectoryURL, error))
-							}
+						.map { _ in true }
+						.flatMapError { error in
+							self._projectEventsObserver.send(value: .skippedInstallingBinaries(project: project, error: error))
+							return SignalProducer(value: false)
```

Comment:
This `flatMapError` could replace the `mapError` after `checkFrameworkCompatibility`. Then you can get rid of `CarthageError.internalError`.